compiler/AST/AggregateType.cpp

```diff
@@ -2224,9 +2226,124 @@ void AggregateType::buildDefaultInitializer() {
+
+  return false;
+}
+
+void AggregateType::buildReaderInitializer() {
+  if (!fUseIOFormatters) return;
+  if (this->getModule()->modTag == MOD_INTERNAL) return;
+
+  if (hasFullyGenericField(this)) return;
+
+  if (builtReaderInit == false &&
+      symbol->hasFlag(FLAG_REF) == false) {
+    SET_LINENO(this);
+    FnSymbol*  fn    = new FnSymbol("init");
+    ArgSymbol* _mt   = new ArgSymbol(INTENT_BLANK, "_mt",  dtMethodToken);
+    ArgSymbol* _this = new ArgSymbol(INTENT_BLANK, "this", this);
+
+    ArgSymbol* reader = new ArgSymbol(INTENT_BLANK, "chpl__reader", dtAny);
+    fn->where = new BlockStmt(new CallExpr("chpl__isFileReader",
+                              new CallExpr(PRIM_TYPEOF, reader)));
+
+    fn->cname = fn->name;
+    fn->_this = _this;
+
+    // Lydia NOTE 06/16/17: I don't think I want to add the
+    //  DEFAULT_CONSTRUCTOR flag to this function, but if I do,
+    // then I will need to do something different in wrappers.cpp.
+    fn->addFlag(FLAG_COMPILER_GENERATED);
+    fn->addFlag(FLAG_LAST_RESORT);
+    fn->addFlag(FLAG_SUPPRESS_LVALUE_ERRORS);
+
+    _this->addFlag(FLAG_ARG_THIS);
+
+    fn->insertFormalAtTail(_mt);
+    fn->insertFormalAtTail(_this);
+    fn->insertFormalAtTail(reader);
+
+
+    if (this->isUnion() == false) {
+      std::set<const char*> names;
+      SymbolMap fieldArgMap;
+
+      if (addSuperArgs(fn, names, fieldArgMap, reader) == true) {
+        // Parent fields before child fields
+        fieldToArg(fn, names, fieldArgMap, reader);
+
+        {
+          CallExpr* getFormatter = new CallExpr(".", reader, new_CStringSymbol("formatter"));
+          CallExpr* readField = new CallExpr("readTypeStart", gMethodToken, getFormatter,
+                                             reader, new SymExpr(this->symbol));
+          fn->insertAtHead(readField);
+        }
+        {
+          CallExpr* getFormatter = new CallExpr(".", reader, new_CStringSymbol("formatter"));
+          CallExpr* readField = new CallExpr("readTypeEnd", gMethodToken, getFormatter,
+                                             reader, new SymExpr(this->symbol));
+          fn->insertAtTail(readField);
+        }
+
+        // Replaces field references with argument references
+        // NOTE: doesn't handle inherited fields yet!
+        update_symbols(fn, &fieldArgMap);
+
+        DefExpr* def = new DefExpr(fn);
+        symbol->defPoint->insertBefore(def);
+
+        fn->setMethod(true);
+        fn->addFlag(FLAG_METHOD_PRIMARY);
+
+        preNormalizeInitMethod(fn);
+
+        normalize(fn);
+
+
+        // BHARSH INIT TODO: Should this be part of normalize(fn)? If we did
+        // that we would emit two use-before-def errors for classes because of
+        // the generated _new function.
+        checkUseBeforeDefs(fn);
+
+        methods.add(fn);
+      } else {
+        USR_FATAL(this, "Unable to generate initializer for type '%s'", this->symbol->name);
+      }
+    } else {
+      // TODO: implement for unions?
```

Comment:
`INT_FATAL` for now here?

Reply:
I'm going to add a check to the beginning of ``buildReaderInitializer`` for unions and return, rather than have this dangling ``else``.